apps/aecore/test/aec_chain_tests.erl

```diff
@@ -1166,6 +1166,129 @@ key_hash_orphaned_chain() ->
+    {ok, {FraudHeader2, FraudHeader1}} = insert_block(MB2),
+
+    ok.
+
+pof_fork_on_micro_block() ->
+    #{ public := PubKey, secret := PrivKey } = enacl:sign_keypair(),
+    PresetAccounts = [{PubKey, 1000}],
+    meck:expect(aec_genesis_block_settings, preset_accounts, 0, PresetAccounts),
+
+    %% Create main chain
+    TxsFun = fun(1) ->
+                     Tx1 = aec_test_utils:sign_tx(make_spend_tx(PubKey, 1, PubKey, 1, 2), PrivKey),
+                     Tx2 = aec_test_utils:sign_tx(make_spend_tx(PubKey, 2, PubKey, 1, 2), PrivKey),
+                     [Tx1, Tx2];
+                (_) ->
+                     []
+             end,
+    [B0, B1, B2, _] = Chain0 = gen_block_chain_with_state_by_target(
+                               PresetAccounts, [?HIGHEST_TARGET_SCI], 1, TxsFun),
+    Chain = [_KB0, _KB1, _MB1, MB2] = blocks_only_chain(Chain0),
+
+    %% Create fork, which starts on a micro-block
+    CommonChain = [B0, B1, B2],
+    Txs = [aec_test_utils:sign_tx(make_spend_tx(PubKey, 2, PubKey, 1, 3), PrivKey)],
+    Fork = aec_test_utils:extend_block_chain_with_micro_blocks(CommonChain, Txs, 1),
+    [_, _, _, MB3] = blocks_only_chain(Fork),
+
+    %% Write initial chain
+    ok = write_blocks_to_chain(Chain),
+
+    %% Write micro-block, and recognize a fraud
+    FraudHeader1 = aec_blocks:to_header(MB2),
+    FraudHeader2 = aec_blocks:to_header(MB3),
+    {ok, {FraudHeader2, FraudHeader1}} = insert_block(MB3),
+
+    ok.
+
+pof_delayed_micro_block() ->
+    #{ public := PubKey, secret := PrivKey } = enacl:sign_keypair(),
+    PresetAccounts = [{PubKey, 1000}],
+    meck:expect(aec_genesis_block_settings, preset_accounts, 0, PresetAccounts),
+
+    %% Create main chain
+    TxsFun = fun(1) ->
+                     Tx1 = aec_test_utils:sign_tx(make_spend_tx(PubKey, 1, PubKey, 1, 2), PrivKey),
+                     Tx2 = aec_test_utils:sign_tx(make_spend_tx(PubKey, 2, PubKey, 1, 2), PrivKey),
+                     [Tx1, Tx2];
+                (_) ->
+                     []
+             end,
+    [B0, B1, B2, _] = Chain0 = gen_block_chain_with_state_by_target(
+                               PresetAccounts, [?HIGHEST_TARGET_SCI], 1, TxsFun),
+    Chain = [KB0, KB1, MB1, MB2] = blocks_only_chain(Chain0),
+
+    %% Create fork, which starts on a micro-block
+    CommonChain = [B0, B1, B2],
+    Txs = [aec_test_utils:sign_tx(make_spend_tx(PubKey, 2, PubKey, 1, 3), PrivKey)],
+    Fork = aec_test_utils:extend_block_chain_with_micro_blocks(CommonChain, Txs, 1),
+    [_, _, _, MB3] = blocks_only_chain(Fork),
+
+    %% Write initial key blocks
+    ok = insert_block(KB0),
+    ok = insert_block(KB1),
+
+    %% Write orphaned micro blocks
+    ok = insert_block(MB2),
+    ok = insert_block(MB3),
+
+    %% Write missing micro block, which recognizes a fraud
+    FraudHeader1 = aec_blocks:to_header(MB2),
+    FraudHeader2 = aec_blocks:to_header(MB3),
+    {ok, {FraudHeader2, FraudHeader1}} = insert_block(MB1),
```

Comment:
This test is failing.

Reply:
1) This is a bug that PoF is not recognized in case of delayed micro-blocks.

In the following state with orphaned `M2` and `M3`:
```
K1 -> __ -> M2
         -> M3
```
When `M1` is inserted a fraud should be recognized for `M2` and `M3` headers, but it is not:
```
K1 -> M1 -> M2
       \
         -> M3
```
This is proven in this test.

To fix this we should look for micro-forks when traversing children of an added block -  https://github.com/aeternity/epoch/blob/pt-155700135-pof-v2-bkp/apps/aecore/src/aec_chain_state.erl#L513. But this does not solve the problem that I describe below 😛 

2. Another bug is when a PoF is recognized after an insert of a micro block in the fork, that is not current top.
Given the following state, with `K4` being the top of the chain with the biggest difficulty (conductor is mining of top of `K4`):
```
K1 -> M1 -> K2 -> M3 -> K3 -> K4
       \
         -> K5 -> M4
```
Then assume `M5` arrives, that is a micro-fork on `K5`, but this does not change a top, i.e.:
```
K1 -> M1 -> K2 -> M3 -> K3 -> K4
       \
         -> K5 -> M4
              \
               -> M5
```

Now `aec_chain_state:insert_block/1` returns `{ok, {M4, M5}}` to the conductor, which is a fraud, but not relevant when mining on a top of K4! This could become relevant when chain on top of either `M4` or `M5` gets the biggest difficulty, and takes over.

To fix this either:
1) `aec_chain_state:insert_block/1` should return a key-block of a generation, in which a fraud has been found; then conductor should compare this returned key-block with top key-block of a generation, on which it is mining,

or

2) new db cache `{key-block, {micro-block-1, micro-block-2}}` should be introduced, which will track where exactly a fraud has been found; this case could be then used by conductor to find if there are any frauds recognized in the current generation

@sennui @hanssv @gorillainduction Does this make any sense?

Reply:
So, both are fine as we discussed, and tagging fraud with key block is required. The new cache solution may be a bit better in regards to managing bloating state of frauds if they happen. Without the cache conductors state would grow.

Reply:
I actually realized that `This could become relevant when chain on top of either M4 or M5 gets the biggest difficulty, and takes over.` is not the case we should worry about. If one of these chains takes over, a fraud should have been recognized already by the miners that posted new key-blocks in these chains. So fix proposed in A) should be sufficient.

I will think about it for a while.